### PR TITLE
[Merged by Bors] - feat(UpperHalfPlane): add positivity extensions

### PR DESCRIPTION
**Pull request by urkud**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean [hunk 1]
```diff
@@ -139,6 +139,32 @@ theorem ne_zero (z : ℍ) : (z : ℂ) ≠ 0 :=
...
+
+open Lean Meta Qq
+
+/-- Extension for the `positivity` tactic: `UpperHalfPlane.im`. -/
+@[positivity UpperHalfPlane.im _]
+def evalUpperHalfPlaneIm : PositivityExt where eval {_ _} _zα _pα e := do
+  let (.app (f : Q(UpperHalfPlane → Real)) (a : Q(UpperHalfPlane))) ← whnfR e
+    | throwError "not UpperHalfPlane.im"
+  guard <|← withDefault <| withNewMCtxDepth <| isDefEq f q(UpperHalfPlane.im)
+  pure (.positive (q(@UpperHalfPlane.im_pos $a) : Expr))
```
Comment (eric-wieser) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:154]: The fully-safe Qq version:
```suggestion
def evalUpperHalfPlaneIm : PositivityExt where eval {u α} _zα _pα e := do
  -- TODO: can't merge the `match`es without lean4#3060
  match u with
  | 0 => match α, e with
    | ~q(ℝ), ~q(UpperHalfPlane.im $a) =>
      assertInstancesCommute
      pure (.positive q(@UpperHalfPlane.im_pos $a))
    | _, _ => throwError "not UpperHalfPlane.im"
  | _ => throwError "not UpperHalfPlane.im"
```

[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:154]: I copied from `ENNReal`. Should I update there too?

File: Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean [hunk 2]
```diff
@@ -139,6 +139,40 @@ theorem ne_zero (z : ℍ) : (z : ℂ) ≠ 0 :=
...
+end UpperHalfPlane
+
+namespace Mathlib.Meta.Positivity
+
+open Lean Meta Qq
+
+/-- Extension for the `positivity` tactic: `UpperHalfPlane.im`. -/
+@[positivity UpperHalfPlane.im _]
+def evalUpperHalfPlaneIm : PositivityExt where eval {u α} _zα _pα e := do
+  -- TODO: can't merge the `match`es without lean4#3060
```
Comment (ocfnash) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:151]: Do I understand correctly that https://github.com/leanprover/lean4/pull/3060 is now in Lean master but we still need to wait for Mathlib to pick up this version of Lean so this PR is currently blocked?

Comment (eric-wieser) [reply Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:151]: No, this PR isn't blocked by that; this TODO comment doesn't need to be addressed immediately, it's just a record of future cleanup.

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Pull request closed without merging**